### PR TITLE
Implement signature verification for self updater

### LINKED ISSUE #1676: Implement gopass self-update
`gopass update` doesn't verify the integrity of downloaded binaries.

If we want to keep this feature we need to implement proper update signing and verification.

### PR DESCRIPTION
**Pull request by dominikschulz**
Fixes #1676

RELEASE_NOTES=[BUGFIX] Add signature verification for updater

Signed-off-by: Dominik Schulz <dominik.schulz@gauner.org>

### CODE REVIEW

File: internal/updater/verify.go [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"strings"
+
+	"github.com/gopasspw/gopass/pkg/debug"
+	"golang.org/x/crypto/openpgp"
+)
+
+var pubkey = []byte(`
```
Comment (AnomalRoil) [inline internal/updater/verify.go:14]: Ideally we could have a secret key in our CI/DI and its pub key would be here.
I guess this correspond to a local key you have on your machine?

File: internal/updater/verify.go [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+			continue
+		}
+		h, err := hex.DecodeString(p[0])
+		if err != nil {
+			return nil, err
+		}
+		return h, nil
+	}
+
+	return nil, fmt.Errorf("hash for file %q not found", filename)
```
Comment (AnomalRoil) [inline internal/updater/verify.go:280]: Ideally the hashsum should be recomputed and compared to that in the hashsum file

File: internal/updater/verify.go (hunk 1 above)
Comment (AnomalRoil) [reply internal/updater/verify.go:14]: Also you could have used a ECDSA key, they are much shorter.

File: internal/updater/download.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+	bar.Done()
+	debug.Log("Transferred %d bytes from %s to %s", count, url, dest)
+	return nil
+}
+
+type setter interface {
+	Set(int64)
+}
+
+type passThru struct {
```
Comment (AnomalRoil) [inline internal/updater/download.go:108]: What for?

File: internal/updater/download.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+		case <-ctx.Done():
+			return backoff.Permanent(errors.New("user aborted"))
+		default:
+		}
+		return download(ctx, dest, url)
+	}, bo)
+}
+
+func download(ctx context.Context, dest, url string) error {
+	fh, err := os.OpenFile(dest, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0755)
```
Comment (AnomalRoil) [inline internal/updater/download.go:76]: You should defer close it, no?

File: internal/updater/github.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+// FetchReleases fetches and returns all releases of gopass from GitHub
+func FetchReleases(pre bool) ([]ghrel.Release, error) {
+	if pre {
+		return ghrel.FetchAllReleases(gitHubOrg, gitHubRepo)
+	}
+	return ghrel.FetchAllStableReleases(gitHubOrg, gitHubRepo)
+}
+
+// LatestRelease fetches and returns the latest gopass release from GitHub
+func LatestRelease(pre bool) (ghrel.Release, error) {
```
Comment (AnomalRoil) [inline internal/updater/github.go:88]: This is available in https://github.com/dominikschulz/github-releases/blob/master/ghrel/releases.go#L111, no?

File: internal/updater/download.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+		name := filepath.Base(header.Name)
+		if header.Typeflag == tar.TypeReg && name == "gopass" {
+			_, err := io.Copy(dfh, tarReader)
+			return errors.Wrapf(err, "Failed to read gopass from tar file")
+		}
+	}
+	return errors.Errorf("file not found in archive")
+}
+
+func tryDownload(ctx context.Context, dest, url string) error {
```
Comment (AnomalRoil) [inline internal/updater/download.go:61]: Isn't the download part done by `ghrel` already?

File: internal/updater/verify.go (hunk 1 above)
Comment (dominikschulz) [reply internal/updater/verify.go:14]: Yes, this is my key for now. That's needs to change eventually, but for implementing and testing that should be OK.

**Comment by AnomalRoil** [comment]
I'm actually wondering whether you shouldn't be adding this capability directly into `ghrel` along with a README.

File: internal/updater/verify.go (hunk 2 above)
Comment (dominikschulz) [reply internal/updater/verify.go:283]: Yes, of course. This wasn't ready for review, yet. Sorry.

File: internal/updater/download.go (hunk 3 above)
Comment (dominikschulz) [reply internal/updater/download.go:80]: This is for the download progressbar.

File: internal/updater/download.go (hunk 6 above)
Comment (dominikschulz) [reply internal/updater/download.go]: No.

File: internal/updater/github.go (hunk 5 above)
Comment (dominikschulz) [reply internal/updater/github.go]: No, I'm ditching that for now.

File: internal/action/version.go [hunk 7]
```diff
@@ -76,30 +77,43 @@ func versionInfo(ctx context.Context, v versioner) string {
 func (s *Action) checkVersion(ctx context.Context, u chan string) {
 	if disabled := os.Getenv("CHECKPOINT_DISABLE"); disabled != "" {
 		u <- ""
+		debug.Log("remote version check disabled by CHECKPOINT_DISABLE")
 		return
 	}
 
-	if strings.HasSuffix(s.version.String(), "+HEAD") || protect.ProtectEnabled {
-		// chan not check version against HEAD or
+	force := os.Getenv("GOPASS_FORCE_CHECK") == "true"
```
Comment (AnomalRoil) [inline internal/action/version.go:84]: `CHECKPOINT_DISABLE` can be any non-null string, but `GOPASS_FORCE_CHECK` must be `"true"`, maybe we should unify?

File: internal/updater/extract.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	switch filepath.Ext(filename) {
+	case ".gz":
+		gzr, err := gzip.NewReader(rd)
+		if err != nil {
+			return err
+		}
+		rd = gzr
+	case ".bz2":
+		rd = bzip2.NewReader(rd)
+	case ".zip":
```
Comment (AnomalRoil) [inline internal/updater/extract.go:36]: So, the updater should exit early on Windows devices then.

Comment (AnomalRoil) [reply internal/updater/extract.go:36]: There's https://golang.org/pkg/archive/zip/ that could handle it.

File: internal/updater/update.go [hunk 9]
```diff
@@ -1,285 +1,101 @@
...
-func download(ctx context.Context, dest, url string) error {
-	fh, err := os.OpenFile(dest, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0755)
-	if err != nil {
-		return err
-	}
-	resp, err := http.Get(url)
-	if err != nil {
+	debug.Log("extracting binary from tarball ...")
+	if err := extractFile(buf, dlFilename, dest); err != nil {
 		return err
```
Comment (AnomalRoil) [inline internal/updater/update.go:95]: So, currently the Windows updates will fail here since `extractFile` doesn't support zip files.

**Review by AnomalRoil — DISMISSED** [review]
LGTM, maybe it'd be best to fix the Windows path?

File: internal/updater/update.go (hunk 9 above)
Comment (dominikschulz) [reply internal/updater/update.go:95]: I guess we should bail out early.

File: internal/updater/extract.go (hunk 8 above)
Comment (dominikschulz) [reply internal/updater/extract.go:36]: Yes, indeed. But I was too lazy do add windows support. I don't think that's much of a problem, but if this is working well on non-Windows OS we can still add Windows support later. Should be fairly easy to plug it in here.

File: internal/action/version.go (hunk 7 above)
Comment (dominikschulz) [reply internal/action/version.go:84]: Good idea.

File: internal/updater/extract.go (hunk 8 above)
Comment (AnomalRoil) [reply internal/updater/extract.go:36]: Sounds good.

**Comment by dominikschulz** [comment]
Windows will now fail early until it's supported.

**Review by AnomalRoil — APPROVED** [review]

**Pull request merged**